refactor: separate payments namespace into payment field and payment channel

foochifa opened the pull request:
## Problem
<!-- What problem are you trying to solve? What issue does this close? -->
Currently we include both payment channel related properties, together with user editable payment properties. Additionally, the payment properties are currently associated with the base form, however, we only intend to push out payment feature for storage mode forms.

This is undesirable as it makes our feature less extensible to future payment channels.

Closes #5872, Closes #5822  

## Solution
<!-- How did you solve the problem? -->
Refactor the schema from payments to `payments_channel` and `payments_field`
Refactor the `payments_channel` and `payments_field` properties to only encrypted form

This has the additional benefit of blocking changes made to `target_account_id` and `publishable_key` from the update payments endpoint. `target_account_id` and `publishable_key` can only be modified through the payment channel's OAuth flow.

This also blocks email mode forms from creating payments.

**Breaking Changes** 
<!-- Does this PR contain any backward incompatible changes? If so, what are they and should there be special considerations for release? -->
- [x] Yes - this PR contains breaking changes
    - This PR will change the schema of the form, the script below will help to migrate existing forms

## Tests
<!-- What tests should be run to confirm functionality? -->

Create a new form.
- [ ] Connect stripe account in settings tab
- [ ] Update payment fields in create tab, and create other form fields
- [ ] Open form, then fill in form and submit a test payment
  
On an existing payment form.
- [ ] Migrate schema by pasting `split-to-payments-field-and-channel.js` into Studio3T and executing it
- [ ] Ensure that you can still submit a test payment on the form
- [ ] Ensure that you can update payment fields and change stripe account

On any payment form.
- [ ] Attempt to update `target_account_id` and `publishable_key` in request body to /:formId/payments endpoint. Request should fail.

On any email form.
- [ ] Attempt to update payment account in payment settings. Should return a http error.
- [ ] Attempt to update payment field from payment tab in create. The drawer and content should not display.

## Deploy Notes

**New scripts**:

- `split-to-payments-field-and-channel.js` : separates the payments property on current forms documents into payments_fields and payments_channel, copy and paste this in studio 3T in the desired DB

foochifa committed:
feat: add payment channel enum to payment.ts

Currently the enum only consists of 'stripe'

foochifa committed:
feat: split FormPayments type to FormPaymentsChannel and FormPaymentsField

foochifa committed:
fix: make target acc id and publishable key optional in payment_channel

To allow for disconnecting and reconnecting stripe acc

foochifa committed:
feat: update payment methods and schema in form.server.model.ts

foochifa committed:
feat: update types for payment and stripe update endpoint

foochifa committed:
feat: update type of storage mode encrypt submission endpoint

foochifa committed:
feat: update types on admin-form payment drawer, settings, response pages

foochifa committed:
feat: update validation of stripe payment account

foochifa committed:
refactor: update frontend public form components

foochifa committed:
refactor: update form server model test cases

foochifa committed:
feat: add script to migrate existing payment forms' schema

foochifa committed:
fix: casing of Stripe payment channel enum

foochifa committed:
fix: remove target_account_id and publishable_key from update payments validation

justynoh commented on shared/constants/form.ts:
```diff
@@ -11,7 +11,7 @@ const PUBLIC_FORM_FIELDS = <const>[
   '_id',
   'responseMode',
-  'payments',
+  'payments_field',
```
should we restrict the `payments_field` to be accessible only by storage mode forms?

justynoh commented on shared/constants/form.ts:
```diff
@@ -21,7 +21,8 @@ export const STORAGE_PUBLIC_FORM_FIELDS = <const>[
 
 const FORM_SETTINGS_FIELDS = <const>[
-  'payments',
+  'payments_channel',
```
same for this, but with `payments_channel`

justynoh commented on shared/types/form/form.ts:
```diff
@@ -104,7 +109,9 @@ export interface FormBase {
   responseMode: FormResponseMode
 
-  payments?: FormPayments
+  payments_channel?: FormPaymentsChannel
+
+  payments_field?: FormPaymentsField
```
somewhat tangentially related to your PR, I wonder if the payments settings should be moved to `StorageFormBase` instead. Initially when we put this here it was because we were planning to support both email and storage. But since we only plan to support Storage mode for payments, maybe we should move these into their appropriate place. A bunch of other stuff will have to change, including the schema definitions, but I think it's worth it. What do you think?

justynoh replied:
This would also make sense with my other thoughts above

foochifa replied:
Yup I agree, didn't know the consideration for email mode was there before haha

In the same note, should I implement that in this PR, or do it in a separate PR to disable email mode admin form payment settings and field?

Cause currently in email mode, we can still connect stripe and edit the payment drawer, which is undesired. #5819

justynoh replied:
yup no worries! this kind of thing we'll just clean up as we go along haha. I think we should do this now, since you're touching the form schema anyway. 

Re: disabling payment settings and field in email mode - we should do all these checks anyway in fe/be, but moving the `payment_field` and `payment_channel` to `StorageModeBase` will give us model-level validation, so even without implementing fe/be checks, the document save should still fail.

foochifa replied:
Yup okay sounds good! Thanks Justyn will make the adjustments!

justynoh reviewed (commented):
generally lgtm, with one question to be discussed

foochifa committed:
fix: shift payments channel and field to storage form settings

foochifa committed:
fix: shift payments_field in to only storage public forms

foochifa committed:
refactor: associate payment properties with encrypted form over formbase

foochifa committed:
refactor: retrieve payment fields at payment tab level

payment field properties will be passed down to payment drawer and content as props

foochifa committed:
refactor: add type check to response page before calling payment field

foochifa committed:
refactor: add typeguard to payment services and controllers

foochifa committed:
refactor: update mutations to use AdminStorageFormDto

foochifa committed:
refactor: shifted payment test cases from base schema to encrypted forms

foochifa committed:
refactor: add typeguard for stripe controller

foochifa committed:
fix: remove reference to payment field in payments settings

This currently blocks the display of payment validation and account id. Which is undesirable as it should still be displayed once account is connected

foochifa commented:
Hey @justynoh have refactored payments into encrypted forms schema. Key modifications:
1. Used typeguards from encrypt submission and form utils for all the payment endpoints. Will throw response mode error if its in email mode.
    a. Maybe I should move checkFormIsEncryptMode method in encrypt submission service to utils? As its now being used outside in multiple other controllers like stripe controller. What do you think?
2. For payment drawer and content, have moved the retrieval of payment field (thru useAdminForm) to payment tab (parent element)
3. Modified the form.server.model (EncryptedFormDocumentSchema), let me know if I have done this correctly, am not familiar with mongo 😅

foochifa committed:
Merge branch 'feat/payment-mvp' into feat/separate-payment-fields-and-payment-channel

foochifa committed:
fix: updated payment properties in formfields and payment preview

justynoh commented on src/app/models/form.server.model.ts:
```diff
@@ -130,8 +133,82 @@ const EncryptedFormSchema = new Schema<IEncryptedFormSchema>({
+      type: Number,
+      validate: {
+        validator: (amount_cents: number) => {
+          if (amount_cents < 50) {
+            return false
+          }
+          if (!Number.isInteger(amount_cents)) {
+            return false
+          }
+          return true
```
suggestions:
`return amount_cents >= 50 && Number.isInteger(amount_cents)`

foochifa replied:
Done 😅 [b26401e](https://github.com/opengovsg/FormSG/pull/5929/commits/b26401eb7e7b6033b4d88632cca33517881a85ad)

justynoh commented on src/app/modules/payments/stripe.service.ts:
```diff
@@ -215,11 +220,21 @@ export const getReceiptURL = (
+                  form.responseMode,
+                ),
+              )
+        })
```
modifying as discussed

foochifa replied:
[d068e95](https://github.com/opengovsg/FormSG/pull/5929/commits/d068e9571e533766319f30179b21af478dfdc388)

justynoh commented on src/types/form.ts:
```diff
@@ -195,19 +196,19 @@ export interface IFormSchema extends IForm, Document, PublicView<PublicForm> {
     accountId,
     publishableKey,
   }: {
-    accountId: FormPayments['target_account_id']
-    publishableKey: FormPayments['publishable_key']
-  }): Promise<T & DeepRequired<Pick<IFormSchema, 'payments'>>>
+    accountId: FormPaymentsChannel['target_account_id']
+    publishableKey: FormPaymentsChannel['publishable_key']
+  }): Promise<T & DeepRequired<Pick<IEncryptedFormSchema, 'payments_channel'>>>
 
   /**
    * Remove payment account ID from the form.
    * @returns updated form
    */
-  removePaymentAccount<T = IFormSchema>(): Promise<T>
+  removePaymentAccount<T = IEncryptedFormSchema>(): Promise<T>
```
As discussed, can we create an issue to move this to `IEncryptedFormSchema`?

foochifa replied:
#5977

justynoh commented on src/types/form.ts:
```diff
@@ -256,6 +257,25 @@ export interface IFormDocument extends IFormSchema {
   responseMode: NonNullable<IFormSchema['responseMode']>
 }
 
+export interface IEncryptedFormDocument extends IEncryptedFormSchema {
+  form_fields: NonNullable<IEncryptedFormSchema['form_fields']>
+  form_logics: NonNullable<IEncryptedFormSchema['form_logics']>
+  permissionList: NonNullable<IEncryptedFormSchema['permissionList']>
+  hasCaptcha: NonNullable<IEncryptedFormSchema['hasCaptcha']>
+  authType: NonNullable<IEncryptedFormSchema['authType']>
+  status: NonNullable<IEncryptedFormSchema['status']>
+  inactiveMessage: NonNullable<IEncryptedFormSchema['inactiveMessage']>
+  // NOTE: Due to the way creating a form works, creating a form without specifying submissionLimit will throw an error.
+  // Hence, using Exclude here over NonNullable.
+  submissionLimit: Exclude<IEncryptedFormSchema['submissionLimit'], undefined>
+  isListed: NonNullable<IEncryptedFormSchema['isListed']>
+  startPage: Required<NonNullable<IEncryptedFormSchema['startPage']>>
+  endPage: Required<NonNullable<IEncryptedFormSchema['endPage']>>
+  webhook: Required<NonNullable<IEncryptedFormSchema['webhook']>>
+  responseMode: NonNullable<IEncryptedFormSchema['responseMode']>
+  publickey: NonNullable<IEncryptedFormSchema['publicKey']>
+}
+
```
extract to generic type, as discussed

foochifa replied:
[ddefa69](https://github.com/opengovsg/FormSG/pull/5929/commits/ddefa69b3ec11e1dd258ae77f49ada250b8d45d4)

justynoh commented on frontend/src/features/public-form/components/FormEndPage/FormEndPageContainer.tsx:
```diff
@@ -63,7 +65,8 @@ export const FormEndPageContainer = ({
   return (
     <Box py={{ base: '1.5rem', md: '2.5rem' }} w="100%">
-      {form?.payments?.enabled ? (
+      {form?.responseMode === FormResponseMode.Encrypt &&
+      form?.payments_field?.enabled ? (
```
at this point should we just make a context variable `isPaymentForm`?

foochifa replied:
#5976

justynoh reviewed (approved):
lgtm, some small things to update but most can be done in separate tickets

foochifa committed:
fix: use retrievefullform in stripe service to utilise shared typeguard

foochifa committed:
refactor: extracted out main form document logic to IFormBaseDocument

foochifa committed:
refactor: made amount_cents validation more concise

justynoh reviewed (approved):
lgtm!
